misc/pytorch_toolkit/noise_suppression/train.py

```diff
@@ -0,0 +1,572 @@
+    epoch_start = int(m.group(1)) if m else 0
+
+    model = models.model_create(args.model_desc)
+
+    size_to_read = model.get_sample_length_ceil(16000 * args.size_to_read)
+
+    printlog("size_to_read {} ".format(size_to_read))
+
+    model = model.to(args.device)
+
+    dataset_train = None
+    # only one process scan folders then share info to other
+    if rank in [-1, 0]:
+        printlog("create train dataset from", args.dns_datasets)
+        dataset_train = dataset.DNSDataset(args.dns_datasets, size_to_read=size_to_read )
+
+    if rank>-1:
+        #lets sync after data creation
+        torch.distributed.barrier()
+        dataset_train = utils.brodcast_data(rank, dataset_train)
+
+    #tune params
+    train = Train(rank, args)
+    train.train(model, dataset_train, epoch_start )
+
+    if rank in [-1, 0]:
+        #save model, eval model
+
+        os.makedirs(args.output_dir, exist_ok=True)
+        model.save_pretrained(args.output_dir)
+
+        # save to onnx
+        model.eval()
+        with torch.no_grad():
+            #get the smallest size
+            size = model.get_sample_length_ceil(1)
+            input_tensor = torch.zeros((1,size), dtype=torch.float, device=args.device)
+
+            outputs = model(input_tensor)
+            state = outputs[2]
+            printlog("state size",sum(t.numel() for t in state),"float params")
+            inputs = (input_tensor, state)
+            input_names = ['input'] + ["inp_state_{:03}".format(i) for i in range(len(state))]
+            output_names = ['output', 'Y']+ ["out_state_{:03}".format(i) for i in range(len(state))]
+
+            #save inputs outputs sizes
+            def rec(n,t):
+                s = list(t.shape)
+                return [
+                    '- name: "{}"'.format(n),
+                    '  shape: '+str(s),
+                    '  layout: ' + ('[N, C, H, W]' if len(s)==4 else '[N, H, W]' if len(s)==3 else '[N, C]'),
+                    '  precision: "FP32"',
+                    '  type: const',
+                    '  float_value: 0'
+                ]
+
+            with open(os.path.join(args.output_dir, "input.yml"), 'wt') as out:
+                out.write("inputs:\n")
+                for n, t in zip(input_names, [input_tensor] + state):
+                    for s in rec(n, t):
+                        out.write("  "+ s + "\n")
+                out.write("outputs:\n")
+                for n, t in zip(output_names, list(outputs[:2]) + list(outputs[2])):
+                    for s in rec(n, t)[:-2]:
+                        out.write("  "+ s + "\n")
+
+            torch.onnx.export(
+                model,
+                inputs,
+                os.path.join(args.output_dir, "model.onnx"),
+                verbose=False,
+                opset_version=11,
+                do_constant_folding=True,
+                input_names=input_names,
+                output_names=output_names)
```

Comment:
I would move export code to the extra file. Is it really needed to export model after every evaluation? Moreover, if user wants to export already trained model, they would have to retrain it again.

Reply:
model is not exported after each evaluation. it is exported right after train is finished. And all trained models saved on onnx. So it is no any reason to export already trained model to onnx again.

Reply:
I meant, after user trained several models, evaluated them and decided to export only the specific model. Anyway, this looks ok also.